### ISSUE TITLE
Iterant Scope uncons hangs on `headOptionL`

### ISSUE THREAD

**Issue by dzajkowski** [monix/monix#1192]
Hi there.
I'm trying to figure out if the following code sample is a bug or a protocol violation:
```scala
val iterant = Iterant[Task].fromReactivePublisher[String](pub)
val head1 = iterant.uncons.headOptionL
val m = head1.flatMap { // <-- (0)
  case Some(Some(res), rest) =>
    val h = rest.headOptionL
    h.runSyncUnsafe(5.minutes) // <-- (1)
    Task(1)
  case _ => Task(0)
}
```
This code will hang at (1).
The direct reason for this is that the IterantFromReactivePublisher runs it's `release` function with `ExitCase.Complete` right around the flatMap call (0). Now the question is - should it?

The `runSyncUnsafe` is just an example, same happens with 
```scala
    val h = rest.headOptionL
    h
```
as h is a `Task`.

a similar piece of code:
```scala
val iterant = Iterant[Task].fromReactivePublisher[String](pub)
val h = iterant.headOptionL.map(_.map(t => (t, iterant.tail)))
h.flatMap { case (Some(hd), tl) =>
  // do something with hd
  tl.headOptionL.map(_.map(t => (t, tl.tail)))
  // one can imagine a def go that recurs over the above structure
}
```
works.
Which is fine but at the same time `tail` has to be computed from scratch on every step which is sub optimal.

To sum this up - if I remove the `out.cancel()` from the release method in `Scope` (which I know is a bad idea) the above code runs fine, but leaks resources.

Hope someone with better understanding of the how `Scope` should behave can give this a look.

Thanks!

**Comment by Avasil** [comment]
I don't know what's causing it off the top of my head because it has been a while since I touched `Iterant` but I will try to fix it for `3.2.2`

**Comment by Avasil** [comment]
What's the use case for `iterant.uncons.headOptionL`? Would it be possible to write it in a different way, let's say:

```scala
    val m = iterant.uncons.mapEval {
      case (Some(res), rest) => rest.headOptionL
      case (None, _) => Task(None)
    }.headOptionL
```

(not sure what's used instead of `rest.headOptionL`)

I'm investigating it and I start to think that calling `release` after `headOptionL` is correct behavior. 

We have: 

```scala
Iterant[F, (Option[A], Iterant[F, A])]
```

If original `Iterant` is a `Scope` then we either leave resource safety responsibilites to the "outer" `Iterant` or the "inner" one.

Currently it's the "outer" and if we call `headOptionL` then it is consumed so finalizers are called.
It is mentioned in ScalaDoc:

```
    * The value returned is wrapped in Iterant to preserve resource safety,
    * and consumption of the rest must not leak outside of use.
```

Although it's not obvious what `use` refers to, I imagine it's a reference of `map` variations to `Resource.use`. 

If we call finalizers when "inner" `Iterant` is done then I feel like it would be quite dangerous - the user could never discard the tail.

Right now I'm thinking we should leave it as is and make the behavior clearer in docs. 
It might be a good idea to signal an error in `IterantFromReactivePublisher` if we try to consume already canceled publisher if it's not against any protocols.
